stdlib/Random/docs/src/index.md

```diff
@@ -146,38 +175,46 @@ julia> rand(Die(4), 3)
-`Random.rand(::AbstractRNG, ::Random.SamplerTrivial{Random.CloseOpen01{T}})`
+Consequently, a method
+```julia
+Sampler(::Type{RNG}, ::Type{T}, n::Repetition) where {RNG<:AbstractRNG,T<:AbstractFloat} =
+    Sampler(RNG, CloseOpen01(T), n)
+```
+is defined to return `SamplerTrivial` with a `Random.CloseOpen01{T}}` type defined for this purpose, which has an appropriate `rand` method defined for it.
 
+#### An optimized sampler with pre-computed data
 
-#### Optimizing generation with cached computation between calls
+Consider a discrete distribution, where numbers `1:n` are drawn with given probabilities that some to one. When many values are needed from this distribution, the fastest method if using an [alias table](https://en.wikipedia.org/wiki/Alias_method). We don't provide the algorithm for building such a table here, but suppose it is available in `make_alias_table(probabilities)` instead, and `draw_number(rng, alias_table)` can be used to draw a random number from it.
 
-When repeatedly generating random values (with the same `rand` parameters), it happens for some types
-that the result of a computation is used for each call. In this case, the computation can be decoupled
-from actually generating the values. This is the case for example with the default implementation for
-`AbstractArray`. Assume that `rand(rng, 1:20)` has to be called repeatedly in a loop: the way to take advantage
-of this decoupling is as follows:
+Suppose that the distribution is described by
+```julia
+struct DiscreteDistribution{V <: AbstractVector}
+    probabilities::V
+end
+```
+and that we *always* want to build an a alias table, regardless of the number of values needed (we learn how to customize this below). The methods
+```julia
+Random.gentype(::Type{<:DiscreteDistribution}) = Int
 
+function Random.Sampler(::AbstractRng, distribution::DiscreteDistribution, ::Repetition)
+    SamplerSimple(disribution, make_alias_table(probabilities))
+end
+```
+should be defined to return a sampler with pre-computed data, then
 ```julia
-rng = MersenneTwister()
-sp = Random.Sampler(rng, 1:20) # or Random.Sampler(MersenneTwister,1:20)
-for x in X
-    n = rand(rng, sp) # similar to n = rand(rng, 1:20)
-    # use n
+function rand(rng::AbstractRNG, sampler::SamplerSimple{<:DiscreteDistribution})
+    draw_number(rng, sampler.data)
 end
 ```
+will be used to draw the values.
+
+#### Custom sampler types
 
-This mechanism is of course used by the default implementation of random array generation (like in `rand(1:20, 10)`).
-In order to implement this decoupling for a custom type, a helper type can be used.
-Going back to our `Die` example: `rand(::Die)` uses random generation from a range, so
-there is an opportunity for this optimization:
+In order to implement this decoupling for a custom type, a custom helper type can be used. Going back to our `Die` example: `rand(::Die)` uses random generation from a range, so there is an opportunity for this optimization:
```

Comment:
I think you should change "In order to implement this decoupling for a custom type", as you already showed how to do the decoupling for a custom type... maybe add "in cases where `SamplerSimple` is not suitable? (but the problem then is that we explain `SamplerSimple` was suitable in our `Die` example!)

Reply:
I tried to clarify this, I hope the improved wording is clear.